### PR TITLE
Alpha/v1.2

### PR DESCRIPTION
**Pull request by danil-lashin**
Temporary pr for code review

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -52,7 +53,6 @@ var Routes = map[string]*rpcserver.RPCFunc{
 	"max_gas":                rpcserver.NewRPCFunc(MaxGas, "height"),
 	"min_gas_price":          rpcserver.NewRPCFunc(MinGasPrice, ""),
 	"genesis":                rpcserver.NewRPCFunc(Genesis, ""),
-	"missed_blocks":          rpcserver.NewRPCFunc(MissedBlocks, "pub_key,height"),
```
[the remark this answers is not in the document]
Comment (klim0v) [reply api/api.go]: ok

File: config/config.go [hunk 2]
```diff
@@ -233,6 +233,9 @@ type BaseConfig struct {
 	// Address to listen for API v2 connections
 	APIv2ListenAddress string `mapstructure:"api_v2_listen_addr"`
 
+	// API v2 Timeout
+	APIv2TimeoutDuration int64 `mapstructure:"api_v2_timeout_duration"`
```
[the remark this answers is not in the document]
Comment (klim0v) [reply config/config.go]: ok

File: config/toml.go [hunk 3]
```diff
@@ -75,10 +75,13 @@ grpc_listen_addr = "{{ .BaseConfig.GRPCListenAddress }}"
...
 api_v2_listen_addr = "{{ .BaseConfig.APIv2ListenAddress }}"
 
+# API v2 Timeout
+api_v2_timeout_duration = "{{ .BaseConfig.APIv2TimeoutDuration }}"
+
 # Sets node to be in validator mode. Disables API, events, history of blocks, indexes, etc. 
 validator_mode = {{ .BaseConfig.ValidatorMode }}
 
-# Sets number of last stated to be saved
+# Sets number of last stated to be saved on disk. If the value is 0, then the KeepEver and KeepRecent strategies are used.
```
[the remark this answers is not in the document]
Comment (klim0v) [reply config/toml.go]: ok

File: core/minter/minter.go [hunk 4]
```diff
@@ -529,14 +529,14 @@ func (app *Blockchain) MissedBlocks(pubKey string, height uint64) (missedBlocks
...
-		cState.Validators.LoadValidators()
+		cState.Validators().LoadValidators()
 		cState.Unlock()
 	}
 
 	cState.RLock()
 	defer cState.RUnlock()
 
-	val := cState.Validators.GetByPublicKey(types.HexToPubkey(pubKey))
+	val := cState.Validators().GetByPublicKey(types.HexToPubkey(pubKey[2:]))
```
[the remark this answers is not in the document]
Comment (klim0v) [reply core/minter/minter.go]: ok

File: core/state/candidates/candidates.go [hunk 5]
```diff
@@ -24,22 +25,50 @@ const (
 	MaxDelegatorsPerCandidate = 1000
 
 	mainPrefix       = 'c'
+	pubKeyIDPrefix   = 'p'
+	blockListPrefix  = 'b'
 	stakesPrefix     = 's'
 	totalStakePrefix = 't'
 	updatesPrefix    = 'u'
+	maxIDPrefix      = 'i'
```
[the remark this answers is not in the document]
Comment (klim0v) [reply core/state/candidates/candidates.go]: ok

File: core/state/candidates/candidates.go [hunk 6]
```diff
@@ -24,22 +25,50 @@ const (
...
+	GetStakes(pubkey types.Pubkey) []*Stake
+}
+
 type Candidates struct {
-	list map[types.Pubkey]*Candidate
+	list map[uint]*Candidate
+
+	isDirty   bool
+	blockList map[types.Pubkey]struct{}
+	pubKeyIDs map[types.Pubkey]uint
```
[the remark this answers is not in the document]
Comment (klim0v) [reply core/state/candidates/candidates.go]: ok

File: core/state/candidates/candidates.go [hunk 7]
```diff
@@ -170,6 +236,13 @@ func (c *Candidates) Create(ownerAddress types.Address, rewardAddress types.Addr
 	c.setToMap(pubkey, candidate)
 }
 
+func (c *Candidates) CreateWithID(ownerAddress, rewardAddress, controlAddress types.Address, pubkey types.Pubkey, commission uint, id uint) {
+	if id != 0 {
```
[the remark this answers is not in the document]
Comment (klim0v) [reply core/state/candidates/candidates.go]: ok

File: core/state/candidates/candidates.go [hunk 8]
```diff
@@ -520,8 +593,30 @@ func (c *Candidates) Exists(pubkey types.Pubkey) bool {
...
+
+func (c *Candidates) existPubKey(pubKey types.Pubkey) bool {
+	if c.pubKeyIDs == nil {
+		return false
+	}
+	_, exists := c.pubKeyIDs[pubKey]
+	return exists
+}
+
+func (c *Candidates) IsBlockPubKey(pubkey *types.Pubkey) bool {
```
[the remark this answers is not in the document]
Comment (klim0v) [reply core/state/candidates/candidates.go]: ok

File: core/state/candidates/candidates.go [hunk 9]
```diff
@@ -520,8 +593,30 @@ func (c *Candidates) Exists(pubkey types.Pubkey) bool {
...
+	if pubkey == nil {
```
[the remark this answers is not in the document]
Comment (klim0v) [reply core/state/candidates/candidates.go]: ok

File: core/state/candidates/candidates.go [hunk 10]
```diff
@@ -684,49 +780,114 @@ func (c *Candidates) GetCandidateOwner(pubkey types.Pubkey) types.Address {
...
+		}
+	}
+
+	return maxID
+}
+
+func (c *Candidates) checkAndSetLoaded() bool {
+	c.lock.RLock()
+	if c.loaded {
+		c.lock.RLock()
```
[the remark this answers is not in the document]
Comment (klim0v) [reply core/state/candidates/candidates.go]: ok

File: core/state/candidates/candidates.go [hunk 11]
```diff
@@ -520,8 +593,30 @@ func (c *Candidates) Exists(pubkey types.Pubkey) bool {
...
+	_, exists := c.pubKeyIDs[pubKey]
+	return exists
+}
+
+func (c *Candidates) IsBlockPubKey(pubkey *types.Pubkey) bool {
+	if pubkey == nil {
+		return false
+	}
+
+	return c.isBlock(*pubkey)
```
[the remark this answers is not in the document]
Comment (klim0v) [reply core/state/candidates/candidates.go]: ok

File: core/state/candidates/candidates.go [hunk 12]
```diff
@@ -991,3 +1184,78 @@ func (c *Candidates) LoadStakesOfCandidate(pubkey types.Pubkey) {
 	candidate.setTmAddress()
 	c.setToMap(candidate.PubKey, candidate)
 }
+
+func (c *Candidates) ChangePubKey(old types.Pubkey, new types.Pubkey) {
+	if old == new {
```
[the remark this answers is not in the document]
Comment (klim0v) [reply core/state/candidates/candidates.go]: ok

File: core/state/state.go [hunk 13]
```diff
@@ -180,8 +305,12 @@ func (s *State) Import(state types.AppState) error {
 	}
 	s.Validators.SetValidators(vals)
 
+	for _, pubkey := range state.BlockListCandidates {
+		s.Candidates.AddToBlockPybKey(pubkey)
```
[the remark this answers is not in the document]
Comment (klim0v) [reply core/state/state.go:309]: ok

File: core/state/candidates/candidates.go [hunk 14]
```diff
@@ -991,3 +1184,78 @@ func (c *Candidates) LoadStakesOfCandidate(pubkey types.Pubkey) {
...
+		return
+	}
+
+	if c.isBlock(new) {
+		panic("Candidate with such public key (" + new.String() + ") exists in block list")
+	}
+
+	c.getFromMap(old).PubKey = new
+	c.setBlockPybKey(old)
```
[the remark this answers is not in the document]
Comment (klim0v) [reply core/state/candidates/candidates.go:1188]: ok

File: core/transaction/encoder/resources.go [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+	}
+
+	return resource
+}
+
+// TxType 0x0E
+
+type EditCandidateDataResource struct {
+	PubKey         string `json:"pub_key"`
+	NewPubKey      string `json:"new_pub_key"`
```
[the remark this answers is not in the document]
Comment (klim0v) [reply core/transaction/encoder/resources.go:275]: ok

File: core/minter/minter.go [hunk 16]
```diff
@@ -169,20 +169,22 @@ func (app *Blockchain) InitChain(req abciTypes.RequestInitChain) abciTypes.Respo
...
-	app.StatisticData().PushStartBlock(statistics.StartRequest{Height: int64(height), Now: time.Now(), HeaderTime: req.Header.Time})
+	app.StatisticData().PushStartBlock(&statistics.StartRequest{Height: int64(height), Now: time.Now(), HeaderTime: req.Header.Time})
 
-	if app.haltHeight > 0 && height >= app.haltHeight {
+	if app.isApplicationHalted(height) {
 		panic(fmt.Sprintf("Application halted at height %d", height))
 	}
 
+	app.lock.Lock()
 	if upgrades.IsUpgradeBlock(height) {
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply core/minter/minter.go:179]: fixed

File: core/minter/minter.go [hunk 17]
```diff
@@ -692,3 +704,49 @@ func getDbOpts(memLimit int) *opt.Options {
 		Filter:                 filter.NewBloomFilter(10),
 	}
 }
+
+func (app *Blockchain) isApplicationHalted(height uint64) bool {
+	if app.haltHeight > 0 && height >= app.haltHeight {
+		return true
+	}
+
+	if height < upgrades.UpgradeBlock4 {
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply core/minter/minter.go:713]: fixed

File: core/minter/minter.go [hunk 18]
```diff
@@ -692,3 +704,49 @@ func getDbOpts(memLimit int) *opt.Options {
...
+		totalPower, totalVotingPower := big.NewInt(0), big.NewInt(0)
+		for _, val := range vals {
+			// skip if candidate is not present
+			if val.IsToDrop() || app.validatorsStatuses[val.GetAddress()] != ValidatorPresent {
+				continue
+			}
+
+			for _, halt := range halts.List {
+				if halt.Pubkey == val.PubKey {
+					totalVotingPower.Add(totalVotingPower, val.GetTotalBipStake())
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply core/minter/minter.go:730]: fixed

File: core/state/bus/coins.go [hunk 19]
```diff
@@ -1,20 +1,32 @@
...
-	Volume  *big.Int
-	Reserve *big.Int
+	ID           types.CoinID
+	Name         string
+	Crr          uint
+	Symbol       types.CoinSymbol
+	Version      types.CoinVersion
+	Volume       *big.Int
+	Reserve      *big.Int
+	OwnerAddress types.Address
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply core/state/bus/coins.go:23]: fixed

File: api/coin_info.go [hunk 20]
```diff
@@ -23,17 +25,25 @@ func CoinInfo(coinSymbol string, height int) (*CoinInfoResponse, error) {
 	cState.RLock()
 	defer cState.RUnlock()
 
-	coin := cState.Coins.GetCoin(types.StrToCoinSymbol(coinSymbol))
+	coin := cState.Coins().GetCoinBySymbol(types.StrToCoinSymbol(coinSymbol))
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply api/coin_info.go:28]: fixed

File: api/coin_info.go [hunk 21]
```diff
@@ -6,12 +6,14 @@ import (
...
-	ReserveBalance string `json:"reserve_balance"`
-	MaxSupply      string `json:"max_supply"`
+	ID             uint32         `json:"id"`
+	Name           string         `json:"name"`
+	Symbol         string         `json:"symbol"`
+	Volume         string         `json:"volume"`
+	Crr            uint           `json:"crr"`
+	ReserveBalance string         `json:"reserve_balance"`
+	MaxSupply      string         `json:"max_supply"`
+	OwnerAddress   *types.Address `json:"owner_address"`
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply api/coin_info.go:16]: fixed

File: api/api.go [hunk 22]
```diff
@@ -44,6 +44,7 @@ var Routes = map[string]*rpcserver.RPCFunc{
 	"events":                 rpcserver.NewRPCFunc(Events, "height"),
 	"net_info":               rpcserver.NewRPCFunc(NetInfo, ""),
 	"coin_info":              rpcserver.NewRPCFunc(CoinInfo, "symbol,height"),
+	"coin_id":                rpcserver.NewRPCFunc(CoinIdInfo, "id,height"),
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply api/api.go:47]: fixed

File: api/addresses.go [hunk 23]
```diff
@@ -5,9 +5,9 @@ import (
 )
 
 type AddressesResponse struct {
-	Address          string            `json:"address"`
-	Balance          map[string]string `json:"balance"`
-	TransactionCount uint64            `json:"transaction_count"`
+	Address          string        `json:"address"`
+	Balance          []BalanceItem `json:"balances"`
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply api/addresses.go:9]: fixed

File: core/transaction/send.go [hunk 24]
```diff
@@ -14,32 +14,40 @@ import (
...
 	To    types.Address
 	Value *big.Int
 }
 
+type Coin struct {
+	ID     uint32 `json:"id"`
+	Symbol string `json:"symbol"`
+}
+
 func (data SendData) MarshalJSON() ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply core/transaction/send.go:27]: fixed

File: core/types/types.go [hunk 25]
```diff
@@ -147,12 +151,52 @@ func (c CoinSymbol) IsBaseCoin() bool {
...
+	return c == GetBaseCoinID()
+}
+
+func (c CoinID) String() string {
+	return strconv.Itoa(int(c))
+}
+
+func (c CoinID) Bytes() []byte {
+	b := make([]byte, 4)
+	binary.BigEndian.PutUint32(b, c.Uint32())
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply core/types/types.go:186]: fixed

File: core/validators/validators.go [hunk 26]
```diff
@@ -1,22 +1,9 @@
...
-	if count > 256 {
-		return 256
-	}
-
-	return int(count)
+	return 64
 }
 
 func GetCandidatesCountForBlock(block uint64) int {
 	return GetValidatorsCountForBlock(block) * 3
```
[the remark this answers is not in the document]
Comment (klim0v) [reply core/validators/validators.go]: ok

File: api/estimate_coin_buy.go [hunk 27]
```diff
@@ -15,7 +15,7 @@ type EstimateCoinBuyResponse struct {
 	Commission string `json:"commission"`
 }
 
-func EstimateCoinBuy(coinToSellString string, coinToBuyString string, valueToBuy *big.Int, height int) (*EstimateCoinBuyResponse, error) {
+func EstimateCoinBuy(coinIdToSell uint64, coinIdToBuy uint64, valueToBuy *big.Int, height int) (*EstimateCoinBuyResponse, error) {
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply api/estimate_coin_buy.go:18]: fixed

File: api/estimate_coin_buy.go [hunk 28]
```diff
@@ -24,29 +24,29 @@ func EstimateCoinBuy(coinToSellString string, coinToBuyString string, valueToBuy
 	cState.RLock()
 	defer cState.RUnlock()
 
-	coinToSell := types.StrToCoinSymbol(coinToSellString)
-	coinToBuy := types.StrToCoinSymbol(coinToBuyString)
+	coinToSell := types.CoinID(coinIdToSell)
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply api/estimate_coin_buy.go:27]: fixed

File: api/estimate_coin_buy.go [hunk 29]
```diff
@@ -24,29 +24,29 @@ func EstimateCoinBuy(coinToSellString string, coinToBuyString string, valueToBuy
...
 		return nil, rpctypes.RPCError{Code: 404, Message: "Coin to buy not exists"}
 	}
 
 	commissionInBaseCoin := big.NewInt(commissions.ConvertTx)
 	commissionInBaseCoin.Mul(commissionInBaseCoin, transaction.CommissionMultiplier)
 	commission := big.NewInt(0).Set(commissionInBaseCoin)
 
-	if coinToSell != types.GetBaseCoin() {
-		coin := cState.Coins.GetCoin(coinToSell)
+	if coinToSell != types.GetBaseCoinID() {
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply api/estimate_coin_buy.go:48]: fixed

File: api/api.go [hunk 30]
```diff
@@ -44,6 +44,7 @@ var Routes = map[string]*rpcserver.RPCFunc{
...
 	"estimate_coin_sell":     rpcserver.NewRPCFunc(EstimateCoinSell, "coin_to_sell,coin_to_buy,value_to_sell,height"),
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply api/api.go:48]: fixed

File: core/state/coins/coins.go [hunk 31]
```diff
@@ -13,224 +12,410 @@ import (
...
+		}
+	}
+
+	recreateCoin.CVersion = lastVersion + 1
+	recreateCoin.isDirty = true
+
+	c.setToMap(recreateCoin.id, recreateCoin)
+	c.markDirty(recreateCoin.id)
+
+	c.Create(newID, recreateCoin.Symbol(), "", volume, crr, reserve, maxSupply, nil)
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply core/state/coins/coins.go:250]: fixed

File: core/transaction/recreate_coin.go [hunk 32]
```diff
@@ -0,0 +1,214 @@
...
+	"github.com/MinterTeam/minter-go-node/core/state"
+	"github.com/MinterTeam/minter-go-node/core/types"
+	"github.com/MinterTeam/minter-go-node/formula"
+	"github.com/tendermint/tendermint/libs/kv"
+	"math/big"
+)
+
+type RecreateCoinData struct {
+	Symbol               types.CoinSymbol
+	InitialAmount        *big.Int
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply core/transaction/recreate_coin.go:17]: fixed

File: core/transaction/recreate_coin.go [hunk 33]
```diff
@@ -0,0 +1,214 @@
...
+	InitialReserve       *big.Int
+	ConstantReserveRatio uint
+	MaxSupply            *big.Int
+}
+
+func (data RecreateCoinData) TotalSpend(tx *Transaction, context *state.CheckState) (TotalSpends, []Conversion, *big.Int, *Response) {
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply core/transaction/recreate_coin.go:23]: fixed

File: core/state/coins/coins.go [hunk 34]
```diff
@@ -13,224 +12,410 @@ import (
...
+	recreateCoin := c.GetCoinBySymbol(symbol)
+	if recreateCoin == nil {
+		panic("coin to recreate does not exists")
+	}
+
+	// update version for recreating coin
+	symbolCoins := c.getBySymbol(symbol)
+
+	// TODO: change array to sorted array by version and get the last one
+	lastVersion := uint16(0)
```
[the remark this answers is not in the document]
Comment (GrKamil) [reply core/state/coins/coins.go:236]: fixed

File: core/minter/minter.go [hunk 35]
```diff
@@ -246,17 +248,18 @@ func (app *Blockchain) BeginBlock(req abciTypes.RequestBeginBlock) abciTypes.Res
 		app.stateDeliver.FrozenFunds.Delete(frozenFunds.Height())
 	}
 
+	if height >= upgrades.UpgradeBlock4 {
```
[the remark this answers is not in the document]
Comment (klim0v) [reply core/minter/minter.go:251]: ok

File: upgrades/blocks.go [hunk 36]
```diff
@@ -3,9 +3,10 @@ package upgrades
 const UpgradeBlock1 = 5000
 const UpgradeBlock2 = 38519
 const UpgradeBlock3 = 109000
+const UpgradeBlock4 = 500000 // TODO: fix this value
```
[the remark this answers is not in the document]
Comment (klim0v) [reply upgrades/blocks.go:6]: ok

**Pull request closed without merging**